### PR TITLE
bpo-32176: Set CO_NOFREE in the code object constructor

### PR DESCRIPTION
**Pull request by ncoghlan**
Previously, CO_NOFREE was set in the compiler, which meant
it could end up being set incorrectly when code objects
were created directly. Setting it in the constructor based
on freevars and cellvars ensures it is always accurate,
regardless of how the code object is defined.

https://bugs.python.org/issue32176

### CODE REVIEW

File: Lib/test/test_code.py [hunk 1]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+        function = List.__getitem__
+        class_ref = function.__closure__[0].cell_contents
+        self.assertIs(class_ref, List)
+
+        # Ensure the code correctly indicates it accesses a free variable
+        self.assertFalse(function.__code__.co_flags & inspect.CO_NOFREE)
+
+        # Ensure the implicit super() call actually works
+        obj = List([1,2,3])
+        self.assertEqual(obj[0], 1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: Does it print to stdout? stdout shouldn't be polluted by tests in non-verbose mode.

How do you know that the injected method actually works instead of the original list method? I suggest to transform the value in `__getitem__`. For example by calling `str()`.

File: Lib/test/test_code.py [hunk 2]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+        # Ensure the closure injection actually worked
+        function = List.__getitem__
+        class_ref = function.__closure__[0].cell_contents
+        self.assertIs(class_ref, List)
+
+        # Ensure the code correctly indicates it accesses a free variable
+        self.assertFalse(function.__code__.co_flags & inspect.CO_NOFREE)
+
+        # Ensure the implicit super() call actually works
+        obj = List([1,2,3])
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: Add spaces after commas.

File: Lib/test/test_code.py [hunk 3]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+
+        def add_foreign_method(cls, f):
+            code = new_code(f.__code__)
+            name = f.__name__
+            defaults = f.__defaults__
+            closure = (f.__closure__ or ()) + create_closure(cls)
+            setattr(cls, name, FunctionType(code, globals(), name, defaults, closure))
+
+        class List(list):
+            def append(self, elem):
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: Are these methods related to the test?

File: Lib/test/test_code.py [hunk 4]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+        # From https://bugs.python.org/issue32176
+        from types import FunctionType, CodeType
+
+        def create_closure(__class__):
+            return (lambda: __class__).__closure__
+
+        def new_code(c_or_f):
+            '''A new code object with a __class__ cell added to freevars'''
+            c = c_or_f.__code__ if isinstance(c_or_f, FunctionType) else c_or_f
+            return CodeType(
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: Seems there are no direct tests for CodeType. I think we need to add them (in different issue).

File: Lib/test/test_code.py [hunk 5]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+    def test_closure_injection(self):
+        # From https://bugs.python.org/issue32176
+        from types import FunctionType, CodeType
+
+        def create_closure(__class__):
+            return (lambda: __class__).__closure__
+
+        def new_code(c_or_f):
+            '''A new code object with a __class__ cell added to freevars'''
+            c = c_or_f.__code__ if isinstance(c_or_f, FunctionType) else c_or_f
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: In this test `new_code()` is called only with the code object.

File: Lib/test/test_code.py [hunk 6]
```diff
@@ -141,6 +147,49 @@ def test_newempty(self):
...
+
+        add_foreign_method(List, __getitem__)
+
+        # Ensure the closure injection actually worked
+        function = List.__getitem__
+        class_ref = function.__closure__[0].cell_contents
+        self.assertIs(class_ref, List)
+
+        # Ensure the code correctly indicates it accesses a free variable
+        self.assertFalse(function.__code__.co_flags & inspect.CO_NOFREE)
```
Comment (serhiy-storchaka) [inline Lib/test/test_code.py]: Add the second argument `hex(function.__code__.co_flags)`.

File: Lib/test/test_code.py (hunk 1 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: Done - replaced the print() with the use of an f-string as the getitem result.

File: Lib/test/test_code.py (hunk 2 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: Done

File: Lib/test/test_code.py (hunk 6 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: Done

File: Lib/test/test_code.py (hunk 3 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: Just copied from the original reproducer in the bug report - replaced with `pass`.

File: Lib/test/test_code.py (hunk 4 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: Yeah, that would probably be a good idea. I did a brief review for other unhandled edge cases in direct calls to the constructor and didn't see any, but I could easily have missed something.

File: Lib/test/test_code.py (hunk 5 above)
Comment (ncoghlan) [reply Lib/test/test_code.py]: I refactored both this and `add_foreign_method` to only handle the cases needed for the test (the original more general versions came from the reported reproducer for the issue)

**Review by ncoghlan — COMMENTED** [review]
Thanks for the review! All suggestions accepted, and I spotted a few other opportunities for cleanups while I was editing.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by ncoghlan** [comment]
@serhiy-storchaka Thanks for the review!
